### PR TITLE
LLE Mapped Buffer: Add unmapping of write buffer, zero-size, and multiple page handling

### PR DESCRIPTION
**Pull request by NarcolepticK**
I had to implement a few more cases in the ipc mapped buffer handling to allow the lle cecd module to run.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4228)

### CODE REVIEW

File: src/core/hle/kernel/ipc.cpp [hunk 1]
```diff
@@ -128,7 +128,11 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
             u32 num_pages =
                 Common::AlignUp(page_offset + size, Memory::PAGE_SIZE) >> Memory::PAGE_BITS;
 
-            ASSERT(num_pages >= 1);
+            // Skip when the size is zero
+            if (size == 0) {
+                i += 1;
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Set the word at the target cmd_buf to 0. I don't know if this is correct behavior but at least it will lead to invalid accesses instead of who-knows-what if the receiver expects to have a buffer there

File: src/core/hle/kernel/ipc.cpp [hunk 2]
```diff
@@ -139,13 +143,23 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+                    ResultCode result = src_process->vm_manager.UnmapRange(
+                        page_start, num_pages * Memory::PAGE_SIZE);
+                    ASSERT(result == RESULT_SUCCESS);
+                } else {
+                    const auto vma_iter = src_process->vm_manager.vma_map.find(source_address);
+                    const auto& vma = vma_iter->second;
+                    const VAddr dest_address = vma.originating_buffer_address;
+
+                    auto buffer = std::make_shared<std::vector<u8>>(size);
+                    Memory::ReadBlock(*src_process, source_address, buffer->data(), size);
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: You can use Memory::CopyBlock for this, there's an overload that takes a source process and a destination process

File: src/core/hle/kernel/ipc.cpp [hunk 3]
```diff
@@ -139,13 +143,23 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
                 // process again because they were (presumably) not modified. This behavior is
                 // consistent with the real kernel.
                 if (permissions == IPC::MappedBufferPermissions::R) {
+                    ResultCode result = src_process->vm_manager.UnmapRange(
+                        page_start, num_pages * Memory::PAGE_SIZE);
+                    ASSERT(result == RESULT_SUCCESS);
+                } else {
+                    const auto vma_iter = src_process->vm_manager.vma_map.find(source_address);
+                    const auto& vma = vma_iter->second;
+                    const VAddr dest_address = vma.originating_buffer_address;
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: I really do not like this originating_buffer_address stuff. You should scan the source cmdbuffer for the corresponding mapped buffer descriptor, they are in order. That's what the real kernel does and what the TODO up there is referencing.

File: src/core/hle/kernel/ipc.cpp (hunk 2 above)
Comment (Subv) [reply src/core/hle/kernel/ipc.cpp]: Mind the target offset when writing the data, it should be written to <Start of original page in client process>+page_offset.

The source_address variable comes from the reply command buffer and already contains the page offset.

File: src/core/hle/kernel/ipc.cpp [hunk 4]
```diff
@@ -128,7 +128,11 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
             u32 num_pages =
                 Common::AlignUp(page_offset + size, Memory::PAGE_SIZE) >> Memory::PAGE_BITS;
 
-            ASSERT(num_pages >= 1);
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Why did you remove this assert?

File: src/core/hle/kernel/ipc.cpp [hunk 5]
```diff
@@ -156,19 +170,13 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
-            // TODO(Subv): Support more than 1 page and aligned page mappings
-            ASSERT_MSG(
-                num_pages == 1 &&
-                    (!IsPageAligned(source_address) || !IsPageAligned(source_address + size)),
-                "MappedBuffers of more than one page or aligned transfers are not implemented");
-
             // TODO(Subv): Perform permission checks.
 
-            // TODO(Subv): Leave a page of Reserved memory before the first page and after the last
+            // TODO(Subv): Leave a page of unmapped memory before the first page and after the last
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: I'm rather sure that the padding page should be Reserved instead of simply unmapped

File: src/core/hle/kernel/ipc.cpp [hunk 6]
```diff
@@ -156,19 +170,13 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
             // page.
 
-            if (!IsPageAligned(source_address) ||
-                (num_pages == 1 && !IsPageAligned(source_address + size))) {
+            if (num_pages == 1 && !IsPageAligned(source_address) &&
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: This condition is wrong.

This code should always run if the source address is not page aligned OR if the buffer ends in a non-page aligned address. Those are actually two separate conditions that have to be split up (basically duplicate the code) now that you support more than 1 page.

File: src/core/hle/kernel/ipc.cpp [hunk 7]
```diff
@@ -195,7 +203,24 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
                                               Kernel::MemoryState::Shared)
                         .Unwrap();
+            } else {
+                auto buffer = std::make_shared<std::vector<u8>>(num_pages * Memory::PAGE_SIZE);
+                Memory::ReadBlock(*src_process, source_address, buffer->data() + page_offset, size);
+
+                // Map the pages into the target process' address space.
+                target_address =
+                    dst_process->vm_manager
+                        .MapMemoryBlockToBase(Memory::IPC_MAPPING_VADDR + Memory::PAGE_SIZE,
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Are you adding Memory::PAGE_SIZE to account for the Reserved padding? You shouldn't do that, you can leave it as a TODO for now

File: src/core/hle/kernel/ipc.cpp [hunk 8]
```diff
@@ -128,76 +139,103 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
-                // Unmap the buffers. Readonly buffers do not need to be copied over to the target
-                // process again because they were (presumably) not modified. This behavior is
-                // consistent with the real kernel.
-                if (permissions == IPC::MappedBufferPermissions::R) {
-                    ResultCode result = src_process->vm_manager.UnmapRange(
-                        page_start, num_pages * Memory::PAGE_SIZE);
-                    ASSERT(result == RESULT_SUCCESS);
+                // Scan the target's command buffer for the matching mapped buffer
+                std::size_t j = dst_untranslated_size;
+                while (j < dst_command_size) {
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: You can make this its own function.

File: src/core/hle/kernel/ipc.cpp [hunk 9]
```diff
@@ -128,76 +139,103 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+                    u32 desc = dst_cmd_buf[j++];
+
+                    if (IPC::GetDescriptorType(desc) == IPC::DescriptorType::MappedBuffer) {
+                        IPC::MappedBufferDescInfo dest_descInfo{desc};
+                        VAddr dest_address = dst_cmd_buf[j];
+
+                        u32 dest_size = static_cast<u32>(dest_descInfo.size);
+                        IPC::MappedBufferPermissions dest_permissions = dest_descInfo.perms;
+
+                        if (permissions == dest_permissions && size == dest_size) {
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Just checking the permissions and size is not enough, i can push two different MappedBuffers with the same permissions and size.

File: src/core/hle/kernel/ipc.cpp [hunk 10]
```diff
@@ -128,76 +139,103 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+                            // process again because they were (presumably) not modified. This
+                            // behavior is consistent with the real kernel.
+                            if (permissions != IPC::MappedBufferPermissions::R) {
+                                // Copy the modified buffer back into the target process
+                                Memory::CopyBlock(*src_process, *dst_process, source_address,
+                                                  dest_address, size);
+                            }
+
+                            // Unmap the Reserved page before the buffer
+                            ResultCode result = src_process->vm_manager.UnmapRange(
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: You should assert that this is actually a reserved page.

File: src/core/hle/kernel/ipc.cpp [hunk 11]
```diff
@@ -128,76 +139,103 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+                                page_start, num_pages * Memory::PAGE_SIZE);
+                            ASSERT(result == RESULT_SUCCESS);
+
+                            // Check if this is the last mapped buffer
+                            VAddr next_reserve = page_start + num_pages * Memory::PAGE_SIZE;
+                            auto& vma =
+                                src_process->vm_manager.FindVMA(next_reserve + Memory::PAGE_SIZE)
+                                    ->second;
+                            if (vma.type == VMAType::Free) {
+                                // Unmap the Reserved page after the last buffer
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Adding the Reserved pages increases complexity by quite a bit, IMO it's fine to just not do it and leave a TODO

File: src/core/hle/kernel/ipc.cpp [hunk 12]
```diff
@@ -128,76 +139,103 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+            target_address = dst_process->vm_manager
+                                 .MapMemoryBlockToBase(
+                                     Memory::IPC_MAPPING_VADDR, Memory::IPC_MAPPING_SIZE, buffer, 0,
+                                     static_cast<u32>(buffer->size()), Kernel::MemoryState::Shared)
+                                 .Unwrap();
 
             cmd_buf[i++] = target_address + page_offset;
+
+            // Check if this is the last mapped buffer
+            if (i < command_size) {
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: Actually, are you sure this happens only on the last MappedBuffer? I seem to remember it maps these guard pages before/after each MappedBuffer

File: src/core/hle/kernel/ipc.cpp (hunk 9 above)
Comment (NarcolepticK) [reply src/core/hle/kernel/ipc.cpp]: Is there something else I can compare for uniqueness? Or is it based on position in the cmd_buf? As the buffers are unmapped I can zero the dst_cmd_buf[j].

File: src/core/hle/kernel/ipc.cpp (hunk 12 above)
Comment (NarcolepticK) [reply src/core/hle/kernel/ipc.cpp]: Are there two guard pages between buffers, or do they share a guard page (as I was assuming)? As the buffers get placed next to each other, there is a reserve page before, between, and at the end.

File: src/core/hle/kernel/ipc.cpp (hunk 9 above)
Comment (Subv) [reply src/core/hle/kernel/ipc.cpp]: I just checked the kernel disassembly: It keeps track of which MappedBuffer was last processed in the current cmdbuf.

The mapped buffers need to appear in the same order in both the source and target cmdbufs, and they have to match exactly, so just keep track of the index position of which mappedbuffer you processed last and start searching from that index

File: src/core/hle/kernel/ipc.cpp (hunk 12 above)
Comment (Subv) [reply src/core/hle/kernel/ipc.cpp]: I'm pretty sure they don't share guard pages

File: src/core/hle/kernel/ipc.cpp [hunk 13]
```diff
@@ -128,76 +140,85 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
-                // consistent with the real kernel.
-                if (permissions == IPC::MappedBufferPermissions::R) {
-                    ResultCode result = src_process->vm_manager.UnmapRange(
-                        page_start, num_pages * Memory::PAGE_SIZE);
-                    ASSERT(result == RESULT_SUCCESS);
+                // Scan the target's command buffer for the matching mapped buffer
+                while (target_index < dst_command_size) {
+                    u32 desc = dst_cmd_buf[target_index++];
+
+                    if (IPC::GetDescriptorType(desc) == IPC::DescriptorType::MappedBuffer) {
```
Comment (Subv) [inline src/core/hle/kernel/ipc.cpp]: You have to handle the other translation types, otherwise you might end up interpreting a Handle as a descriptor. Also please move this to its own function, it's big enough.

Other than that, LGTM.

File: src/core/hle/kernel/ipc.cpp [hunk 14]
```diff
@@ -128,76 +140,85 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+            if (size == 0) {
+                cmd_buf[i++] = 0;
+                break;
+            }
             ASSERT(num_pages >= 1);
 
             if (reply) {
-                // TODO(Subv): Scan the target's command buffer to make sure that there was a
-                // MappedBuffer descriptor in the original request. The real kernel panics if you
-                // try to reply with an unsolicited MappedBuffer.
```
Comment (B3n30) [inline src/core/hle/kernel/ipc.cpp]: The real kernel panics if you try to reply with an unsolicited MappedBuffer.

Shouldn’t this still happen in Citra too?

Comment (Subv) [reply src/core/hle/kernel/ipc.cpp]: That's a good point, why did you remove that part of the comment?

Comment (NarcolepticK) [reply src/core/hle/kernel/ipc.cpp]: Copy/paste error. Thanks for noticing. We've been sick. Will fix up soon.

File: src/core/hle/kernel/ipc.cpp [hunk 15]
```diff
@@ -128,76 +140,85 @@ ResultCode TranslateCommandBuffer(SharedPtr<Thread> src_thread, SharedPtr<Thread
...
+                        // Unmap the buffer and guard pages from the source process
+                        ResultCode result = src_process->vm_manager.UnmapRange(
+                            page_start - Memory::PAGE_SIZE, (num_pages + 2) * Memory::PAGE_SIZE);
+                        ASSERT(result == RESULT_SUCCESS);
+
+                        target_index += 1;
+                        break;
+                    }
+
+                    target_index += 1;
```
Comment (wwylele) [inline src/core/hle/kernel/ipc.cpp]: `target_index` is increased twice per iteration: here and at the beginning `u32 desc = dst_cmd_buf[target_index++];`. I guess this is not what you wanted

**Comment by B3n30** [comment]
@NarcolepticK Any plans to address the review comments?
